### ISSUE TITLE
Changes to macros are ignored

### ISSUE THREAD

**Issue by cstjean** [timholy/Revise.jl#20]
I've been poking at Revise and found an issue with macros, that you're probably already aware of:

```julia
module B
macro some_val()
    4   # change here
end

f() = @some_val()   # f() doesn't get updated
end
```

**Comment by timholy** [comment]
I don't think it's Revise-specific:
```julia
julia> module B
       macro some_val()
           4   # change here
       end

       f() = @some_val()   # f() doesn't get updated
       end
B

julia> B.f()
4

julia> eval(B, quote
       macro some_val()
           5
       end
       end)
WARNING: Method definition @some_val() in module B at REPL[1]:3 overwritten at REPL[3]:3.
@some_val (macro with 1 method)

julia> B.f()
4
```

@vtjnash, I presume this is a known limitation of the fix of 265? Should I file an issue or not?

**Comment by vtjnash** [comment]
No, this isn't a bug, that Julia is not a pure functional language.

**Comment by timholy** [comment]
Understood. I'll add a caveat to the README.

**Comment by cstjean** [comment]
> No, this isn't a bug, that Julia is not a pure functional language.

That macros are expanded at "parse-time" (or shortly thereafter) instead of compile-time is a Lisp hold-over, no? 

> Understood. I'll add a caveat to the README.

If there was a function to force reloading every function in the module, it might be a good work-around. In ClobberingReload, any change to the file reloads the entire module, which is wasteful, but also solves this issue in the single-module case (and for multiple modules, it's possible to specify dependencies explicitly).

**Comment by vtjnash** [comment]
That said, Revise.jl can track any additional information it would like, to (partially) cover additional cases. So it could keep track of macros and for direct redefinition know which uses to re-expand.

**Comment by vtjnash** [comment]
> That macros are expanded at "parse-time" (or shortly thereafter) instead of compile-time is a Lisp hold-over, no?

No, Julia (like lisp) expand them at expansion time. That's just the definition of what it means to be a macro. A function that is expanded a runtime is known simply as "a function".

**Comment by timholy** [comment]
> So it could keep track of macros and for direct redefinition know which uses to re-expand.

True. I'll reopen. I don't plan on implementing this anytime soon, but for macro definitions that occur after Revise is loaded (i.e., not ones in Base), this doesn't sound too hard.

**Comment by timholy** [comment]
This was partially fixed by #41. The other suggestion, that Revise manually figure out which methods are using which macros, is less desirable now that Revise doesn't parse the source code of most packages until necessary (due to the source-cache). That's pretty important to keep Revise's overhead low, so I'm going to close this as a wontfix.

Any future improvements will have to come in Julia's Base, adding appropriate backedges.
